groovy-language-server/src/main/java/com/palantir/ls/util/SourceWriter.java

```diff
@@ -0,0 +1,191 @@
+                        StandardCharsets.UTF_8.toString()));
+        File tempFile = File.createTempFile("tempdestination", ".tmp");
+        BufferedWriter output =
+                new BufferedWriter(new OutputStreamWriter(new FileOutputStream(tempFile.getAbsolutePath(), false),
+                        StandardCharsets.UTF_8));
+
+        int changeIdx = 0;
+        boolean endOfFile = false;
+        String line = file.readLine();
+        int lineNum = 0;
+        if (line == null) {
+            endOfFile = true;
+        }
+        int lastColumn = 0;
+
+        for (; !endOfFile && changeIdx < sortedChanges.size(); changeIdx++) {
+            Position start = sortedChanges.get(changeIdx).getRange().getStart();
+            Position end = sortedChanges.get(changeIdx).getRange().getEnd();
+
+            // Find the line where this change starts.
+            while (start.getLine() != lineNum) {
+                // Append this line which is not affected by a range.
+                output.write(line.substring(Math.min(lastColumn, line.length())));
+                output.newLine();
+                lastColumn = 0;
+                line = file.readLine();
+                ++lineNum;
+                if (line == null) {
+                    endOfFile = true;
+                    break;
+                }
+            }
+
+            if (endOfFile) {
+                break;
+            }
+
+            // Handle this change
+            // Add everything before this range starting at the last column.
+            output.write(
+                    line.substring(Math.min(lastColumn, line.length()), Math.min(start.getCharacter(), line.length())));
+            output.write(sortedChanges.get(changeIdx).getText());
+
+            // Advance the file line buffer to where the range ends, ignoring all those lines.
+            while (end.getLine() != lineNum) {
+                line = file.readLine();
+                ++lineNum;
+                if (line == null) {
+                    endOfFile = true;
+                    break;
+                }
+            }
+
+            // Set the where the next line should start, which is where the range ends since it is exclusive.
+            lastColumn = end.getCharacter();
+        }
+
+        // Add the remaining file lines that are not affected by ranges
+        appendRemainingFile(file, line, output, lastColumn);
+
+        file.close();
+
+        // Add the remaining changes that are out of range of this file
+        appendRemainingRanges(sortedChanges, changeIdx, output);
+
+        output.close();
+        FileUtils.copyFile(tempFile, destination.toFile());
+    }
+
+    private synchronized void appendRemainingFile(BufferedReader file, String currentLine, BufferedWriter output,
+            int lastColumn)
+            throws IOException {
+        String line = currentLine;
```

Comment:
refactor this part into (didn't try to compile this):

output.write(currentline.substring(Math.min(lastColumn, line.length)));
output.newLine()
while ((line = file.read()) != null) {
    output.write(line.substring(lastColumn))
}